### ISSUE TITLE
Can surface loss be understood in following not efficient but simple way?

### ISSUE THREAD

**Issue by JunMa11** [LIVIAETS/boundary-loss#5]
Thanks for sharing the code.
I very enjoy your nice idea and reading your paper.
But I'm confused about the implementation of [surface loss.](https://github.com/LIVIAETS/surface-loss/blob/108bd9892adca476e6cdf424124bc6268707498e/losses.py#L74)
In section 2 of the paper, it said "boundary loss in Eq. (5) is the sum of linear functions of the regional softmax probability outputs of the network".

Without regard to the efficiency of the code, can `surfaceloss` be understood in following way?
```python
# inputs (2D/3D):
    # ground_truth
    # cnn_prediction: softmax output
# outputs: surface_loss

# precompute the level set function
gt_onehot = class2one_hot(ground_truth)
gt_dist = one_hot2dist(gt_onehot)

def surface_loss(CNN_prediction, gt_dist):
    cnn_pred_onehot = class2one_hot(CNN_prediction)
    loss = np.mean(cnn_pred_onehot, gt_dist)
    return loss
```

Looking forward to your reply.
Best regards,
Edward

**Comment by HKervadec** [comment]
Should be `np.mean(cnn_pred_onehot * gt_dist)` (probs a typo), but yes this looks like a correct simplification of what we do. 

The code is a little bit more complex, because we select which classes should be supervised with the surface loss. 

`one_hot2dist` will compute one distance map for each class. For a problem with `4` classes, `0` being the background, maybe you want to compute the surface loss only wrt the distance maps of classes `1`, `2` and `3`. Or maybe all of them, or maybe only for the class `2`.  So `self.idc` is the list of classes that you should supervise with this loss (`self.idc = [1]` in the paper experiments as we sticked to binary problems).

Then, I use `einsum` so I can easily remember the shape of my tensor when I come back after a while (bcwh standing for batch,class,width,height). It will also explode if I feed by accident a tensor with less or more axises than 4, which makes debugging easier. 

Let me know if some parts remain difficult to understand.

**Comment by JunMa11** [comment]
Hi @HKervadec , thanks for your quick response very much.
